domains/auth/Auth.API/API/Controllers/LoginController.cs

```diff
@@ -28,4 +51,92 @@ public LoginController(IOidcService oidcService)
+    [Route("/oidc/login/callback")]
+    public async Task<ActionResult<NextStep>> CallbackAsync(
+        OidcCallbackParams oidcCallbackParams,
+        [FromServices] ICryptographyFactory stateCryptography,
+        [FromServices] ICryptographyFactory IdtokenCryptography
+    )
+    {
+        AuthState authState;
+        IdTokenInfo oidcIdToken;
+        UserInfoToken oidcUserInfoToken;
+        OidcTokenResponse oidcToken;
+
+        try
+        {
+            authState = stateCryptography.StateCryptography().Decrypt<AuthState>(oidcCallbackParams.State) ?? throw new InvalidOperationException();
+
+        }
+        catch (Exception)
+        {
+            return BadRequest();
+        }
+
+        var validationResult = await validator.ValidateAsync(oidcCallbackParams);
+
+        if (!validationResult.IsValid)
+        {
+            var redirectlocation = oidcService.OnOidcFlowFailed(authState, oidcCallbackParams);
+            return Redirect(redirectlocation.NextUrl);
+        }
+
+        try
+        {
+            oidcToken = await oidcService.FetchToken(oidcCallbackParams.Code);
+            oidcIdToken = DeserializeToken<IdTokenInfo>(oidcToken.IdToken);
+            oidcUserInfoToken = DeserializeToken<UserInfoToken>(oidcToken.UserinfoToken);
+        }
+        catch (Exception)
+        {
+            var redirectUrl = oidcService.BuildFailureUrl(authState, AuthError.FailedToCommunicateWithIdentityProvider);
+            return Redirect(redirectUrl.NextUrl);
+        }
+
+        if (oidcUserInfoToken.IsPrivate)
+        {
+            await oidcService.Logout(authState.IdToken);
+            var redirectUrl = oidcService.BuildFailureUrl(authState, AuthError.PrivateUsersNotAllowedToLogin);
+            return Redirect(redirectUrl.NextUrl);
+        }
+
+        var user = await userStorage.UserByOidcReferences(oidcIdToken.Sub, oidcIdToken.Idp);
+        var company = (authState.Tin != null) ? await companyStorage.CompanyByTin(authState.Tin) : null;
+
+        if (user == null)
+        {
+            var newAuthState = new AuthState
+            {
+                FeUrl = authState.FeUrl,
+                ReturnUrl = authState.ReturnUrl,
+                TermsAccepted = authState.TermsAccepted,
+                IdToken = IdtokenCryptography.IdTokenCryptography().Encrypt(oidcToken.IdToken),
+                Tin = oidcUserInfoToken.NemidCvr,
+                IdentityProvider = oidcIdToken.Idp,
+                ExternalSubject = oidcIdToken.Sub,
+                CustomerType = authState.CustomerType
+            };
+
+            var redirectUrlTerms = new NextStep
+            {
+                NextUrl = newAuthState.FeUrl + $"/terms?state={stateCryptography.StateCryptography().Encrypt(newAuthState)}"
+            };
+            return Redirect(redirectUrlTerms.NextUrl);
+        }
+        else
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    internal T DeserializeToken<T>(string token)
```

Comment:
Would it make sense to indicate that this is JWT specific tokens?

And does it make sense that the login controller now has the ownership of deserializing tokens?
It looks like the intention is to deserialize to one of our token classes in the `Model` layer. Maybe we can create an abstract `token` class that handle the deserialization?